### PR TITLE
Refactor lerax-square example: use GameLoop and InputHandler

### LINKED ISSUE #19: Refactor lerax-square example to use InputHandler and GameLoop
Make it more brief, delete useless code.

### PR DESCRIPTION
**Pull request by ryukinix**
Closes #19

CHANGELOG:
- Use GameLoop instead and remove the following methods:
  - constructor
  - destructor
  - run
- Use InputHandler and simplify Game::event method

From 226 lines, now that example has only 152. We can reduce a few
lines maybe using ColliderScreen too. Maybe.

### CODE REVIEW

**Comment by ryukinix** [comment]
Next week I'll merge this. Feel free to review. @wesleycsj @rafaelcn

File: examples/lerax-square.cpp [hunk 1]
```diff
@@ -1,16 +1,30 @@
...
 #include "math/Vector2.hpp"
+#include "GameLoop.hpp"
+#include "InputHandler.hpp"
 
 #define SCREEN_WIDTH 800
 #define SCREEN_HEIGHT 600
 
 using namespace std;
 
+// keys to InputHandler observe
```
Comment (rafaelcn) [inline examples/lerax-square.cpp:17]: Wouldn't it be better to have this in the EventHandler namespace as a class enum? This would provide the safety (with types) and also be better placed inside a piece of object that is responsible for events.

File: examples/lerax-square.cpp [hunk 2]
```diff
@@ -1,16 +1,30 @@
...
+    {"quit", SDLK_ESCAPE},
+    {"up", SDLK_UP},
+    {"left", SDLK_LEFT},
+    {"right", SDLK_RIGHT},
+    {"down", SDLK_DOWN},
+    {"velocity_up", SDLK_w},
+    {"velocity_down", SDLK_s},
+};
+
 struct KeyboardState {
```
Comment (rafaelcn) [inline examples/lerax-square.cpp:28]: Why is this here still?

File: examples/lerax-square.cpp [hunk 3]
```diff
@@ -53,75 +67,35 @@ struct Player {
...
-class Game {
+
+class Game: public GameLoop {
 private:
-    SDL_Window *window = nullptr;
-    SDL_Renderer *renderer = nullptr;
-    const char *title = "Square Moving";
-    bool running = true;
-    struct Player player;
+    using GameLoop::GameLoop;
```
Comment (rafaelcn) [inline examples/lerax-square.cpp:73]: I would rather prefer you to place this expression inside a function where it's used. It's quite cumbersome to read this inside a class declaration let alone interpret this in the scope of the class.

File: examples/lerax-square.cpp [hunk 4]
```diff
@@ -53,75 +67,35 @@ struct Player {
...
-                }
-                break;
-            }
-        }
+    InputHandler input_handler = InputHandler(input_config);
+
+    void event() {
+        static SDL_Event e;
+        input_handler.process(e);
+        keyboard.up = input_handler.read("up");
```
Comment (rafaelcn) [inline examples/lerax-square.cpp:81]: I like this! But would the semantic meaning of this function would be better read as `act` instead of `read`?

File: examples/lerax-square.cpp [hunk 5]
```diff
@@ -53,75 +67,35 @@ struct Player {
...
+        keyboard.velocity_up = input_handler.read("velocity_up");
+        keyboard.velocity_down = input_handler.read("velocity_down");
+        running = !(input_handler.read(SDL_QUIT) || input_handler.read("quit"));
     }
 
     void update(float dt) {
         if (keyboard.velocity_up) {
             player.velocity += 50;
-            keyboard.velocity_up = false;
+            input_handler.write("velocity_up", false);
```
Comment (rafaelcn) [inline examples/lerax-square.cpp:93]: Yep, referring to the previous comment about this `InputHandler` class I can see why you have done it that way. Well, events are either read or written inside a buffer, be that buffer a cyclic one or not. This is quite good for now.

**Review by rafaelcn — DISMISSED** [review]

File: examples/lerax-square.cpp (hunk 1 above)
Comment (ryukinix) [reply examples/lerax-square.cpp:17]: Why and how? That is a specific thing for this example. What you think should be in the enum?

File: examples/lerax-square.cpp (hunk 2 above)
Comment (ryukinix) [reply examples/lerax-square.cpp:28]: This is just a collection of state that InputHandler will save the result on it. I'm not providing Player direct access to InputHandler, rather that I'm processing the events on `Game::event` and saving on KeyboardState.

How do you think I can improve it?

File: examples/lerax-square.cpp (hunk 4 above)
Comment (ryukinix) [reply examples/lerax-square.cpp:81]: Why act?

File: examples/lerax-square.cpp (hunk 3 above)
Comment (ryukinix) [reply examples/lerax-square.cpp:73]: I was trying to kill a annoying compiling error. I can move that to `main` if that still fixes it, but I'm not sure. BTW, why the constructor is not inherited by default?

File: examples/lerax-square.cpp (hunk 1 above)
Comment (wesleycsj) [reply examples/lerax-square.cpp:17]: This enum should die or be reallocated to internals of hazard-memory. At this time it could be done like that, so we gonna discuss to separate the examples PR from library design PRs in the future.

**Comment by wesleycsj** [comment]
Everything seems default and ok for me.

File: examples/lerax-square.cpp (hunk 3 above)
Comment (wesleycsj) [reply examples/lerax-square.cpp:73]: I already have suggest to clean the syntax in future versions, instead of inherit a class, only declare the common functions like: draw(), update(), keyevent() and etc.Limiting unnecessary back-end information to the developer and make syntax more clean and fast to code.

Comment (wesleycsj) [reply examples/lerax-square.cpp:73]: I already have suggest to clean the syntax in future versions, instead of inherit a class, only declare the common functions like: draw(), update(), keyevent() and etc.Limiting unnecessary back-end information to the developer and make syntax more clean and fast to code.

File: examples/lerax-square.cpp (hunk 2 above)
Comment (wesleycsj) [reply examples/lerax-square.cpp:28]: Let's discuss everything which is about hazard-memory lib in the issue space.

File: examples/lerax-square.cpp (hunk 4 above)
Comment (rafaelcn) [reply examples/lerax-square.cpp:81]: Because the input handler would act over the "up" event instead of reading the event (?).

**Comment by ryukinix** [comment]
Can someone approve that MR? @rafaelcn @wesleycsj

File: examples/lerax-square.cpp (hunk 4 above)
Comment (ryukinix) [reply examples/lerax-square.cpp:81]: Actually it's just read the state map of a given key

File: examples/lerax-square.cpp (hunk 3 above)
Comment (ryukinix) [reply examples/lerax-square.cpp:73]: That can be made including a new header called `Game.hpp` containing only declarations

**Comment by wesleycsj** [comment]
Merging blocked for me. You should resolve rafnu request changes first.

**Comment by ryukinix** [comment]
I'm request to accept the code review, not merge it @wesleycsj. I mean:

Files Changed > Review Changes > Approve > Submit Review

I will not merge now because I need to rebase against master branch.

File: examples/lerax-square.cpp (hunk 4 above)
Comment (wesleycsj) [reply examples/lerax-square.cpp:81]: I think the intention wasn't propagate events but only get a key, sure?

**Review by wesleycsj — APPROVED** [review]
Everything ok, for now.

File: examples/lerax-square.cpp (hunk 3 above)
Comment (ryukinix) [reply examples/lerax-square.cpp:73]: We should create a issue only to get rid of that particular problem of `GameLoop::GameLoop`

File: examples/lerax-square.cpp (hunk 4 above)
Comment (ryukinix) [reply examples/lerax-square.cpp:81]: Yes, maybe using `get` and `set` can be more appropriate in this case.

File: examples/lerax-square.cpp (hunk 1 above)
Comment (ryukinix) [reply examples/lerax-square.cpp:17]: It's not clear to me what would be in the enum. I think this should be better addressed to a particular issue.
